Code framework with examples

wollip opened the pull request:
There is a picture of all the classes we need and the functions within them.

There are some templates so your class will have the necessary functions

There is also an example of the Tape follower in this framework.

TODO change config for DEVICE() from 0 to 1

Commit:
Add Class template

Commit:
Add hardware interface

Commit:
Add Tape Follower

Add update to tape

Commit:
Add correct code heiarchy picture

minor bug fixes

fb39ca4 commented:
You should just use #pragma once. It's not standards compliant but #ifndef guards are annoying to write. ¯\_(ツ)_/¯

wollip commented on include/configs/configs.h:
```diff
@@ -0,0 +1,27 @@
+#ifndef WALKERVILLE_ROBOT_CONFIGS_H
+#define WALKERVILLE_ROBOT_CONFIGs_H
```
Fix typo please

wollip commented on include/configs/configs.h:
```diff
@@ -0,0 +1,27 @@
+// debug and generation use
+#define USE_UPDATE() 1
+#define DEBUG() 1
+#define DEVICE() 0
```
Change to 1 and test on robot

wollip commented on src/sequences/tape.cpp:
```diff
@@ -0,0 +1,103 @@
+  return gain_t_*(kp_ + ki_ +  kd_);
+}
+
+#if USE_UPDATE()
```
change the indentation for this seciton

wollip commented on include/configs/configs.h:
```diff
@@ -0,0 +1,27 @@
+#ifndef WALKERVILLE_ROBOT_CONFIGS_H
```
might be nice to have both the header guard and `#pragma once`

wollip commented on src/hardware/driver.cpp:
```diff
@@ -0,0 +1,27 @@
+#else  // macro !DEVICE()
+void Driver::sendMotorCommand(int velocity, int command)
+{
+  std::cout << "l_motor: " <<  velocity - command << std::endl;
```
Add TODO here for the future if we want the robot simulator

wollip commented on src/hardware/ir.cpp:
```diff
@@ -0,0 +1,42 @@
+  l_val_ = analogRead(L_SENSOR_);
+  r_val_ = analogRead(R_SENSOR_);
+#else 
+  l_val_ = 800;
```
add TODO here for the future

wollip commented on src/hardware/ir.cpp:
```diff
@@ -0,0 +1,42 @@
+#include "hardware/ir.h"
```
`#include "arduino.h"` if DEVICE()

wollip commented on src/hardware/driver.cpp:
```diff
@@ -0,0 +1,27 @@
+#include "hardware/driver.h"
+
```
include the enph stuff it is DEVICE()

wollip commented on src/sequences/tape.cpp:
```diff
@@ -0,0 +1,103 @@
+{
+  error_ = ir_.getTapeError();
+  command_ = computeCommand(error_, 100);
+#if DEBUG() 
```
might have to figure out how we want to set the debugs because we have a limited amount of screen space

wollip commented on src/sequences/tape.cpp:
```diff
@@ -0,0 +1,103 @@
+void update()
+{
+#if DEVICE()
+  while (!startbutton())
```
might want to redo this entire menu system.

wollip reviewed (commented):
Looks okay, but needs some changes code wise

Commit:
Add script to convert cpp directory to arduino

wollip commented on include/configs.h:
```diff
@@ -0,0 +1,27 @@
+
+#define R_MOTOR() 0
+#define L_MOTOR() 1
+#define L_SENSOR() 2
```
R_SENSOR then L_SENSOR

wollip commented on include/hardware/ir.h:
```diff
@@ -0,0 +1,35 @@
+  int getTapeError();
+  bool tenKHZ();
+  int strength();
+
```
Add an `update` that can be called from `main.cpp`

Commit:
Add update to Ir;

minor comment changes

convert clrf to lf

Commit:
Add batch script

Fix error bug

update configs to tape follow

Change script to link instead of copy
